clients/drcachesim/tools/invariant_checker.cpp

```diff
@@ -469,17 +495,18 @@ invariant_checker_t::parallel_shard_memref(void *shard_data, const memref_t &mem
                     TESTANY(OFFLINE_FILE_TYPE_FILTERED | OFFLINE_FILE_TYPE_IFILTERED,
                             shard->file_type_),
                 "Signal handler return point incorrect");
-            // We assume paired signal entry-exit (so no longjmp and no rseq
-            // inside signal handlers).
-            shard->prev_xfer_int_pc_.pop();
-            shard->pre_signal_instr_.pop();
         }
+        // It is a little inefficient to replace the top instr on the stack everytime.
+        // But we cannot perform this book-keeping using prev_instr_ on a
+        // TRACE_MARKER_TYPE_KERNEL_EVENT marker. If there was no instruction between
+        // two nested signals, we do not want to record any pre-signal instr for the
```

Comment:
But we do want the pre-signal instr for an instr prior to two consecutive signals at the same scope.  What we want is a prev_instr stack so we get the prev_instr for our scope -- but isn't that what shard->pre_signal_instr is already?  I think I'm confused and not understanding this right: not understanding this replace_top.

Reply:
> But we do want the pre-signal instr for an instr prior to two consecutive signals at the same scope

That's right. We do do that here. See comment in the "Multiple nested signals without any intervening instr between the nested" test.

> but isn't that what shard->pre_signal_instr is already?

Yes.

> not understanding this replace_top

When we see a new instr, that instr becomes the last instr seen in the current scope. That's what the replacement means.

Reply:
What we want is shard->cur_scope->prev_instr.  Instead of popping and pushing to store prev_instr, if we do the compound data structure below we can more simply store into the cur scope (== top of scope/context stack)'s field for prev_instr.

Should we go further and get rid of shard->prev_instr in general and always use shard->cur_scope->prev_instr everywhere?

Reply:
I switched to using a separate var `last_instr_in_cur_context_` that keeps getting updated.

Reply:
> Should we go further and get rid of shard->prev_instr in general and always use shard->cur_scope->prev_instr everywhere?

That'd get a little complex. `shard->cur_scope->prev_instr ` (or shard->signal_stack_.top().pre_signal_instr in the code now) may not be set. It can be `{ }` (to denote absence of any pre-signal instr in that context). So we'll need to search the whole signal_stack_ from the top to find the first entry with a valid pre_signal_instr.